arch/risc-v/src/common/espressif/esp_mcpwm.c

```diff
@@ -0,0 +1,682 @@
+static struct mcpwm_cap_channel_lowerhalf_s mcpwm_cap_ch2_lowerhalf =
+{
+  .ops        = &mcpwm_cap_ops,
+  .common     = &mcpwm_common,
+  .data       = &event_data_ch2,
+  .channel_id = MCPWM_CAP_CHANNEL_2,
+  .ready      = false,
+};
+#endif
+#endif  /* CONFIG_ESP_MCPWM_CAPTURE */
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp_capture_start
+ *
+ * Description:
+ *   This function is a requirement of the upper-half driver. When called,
+ *   enables the capture channel, interruption routine, sets the positive
+ *   edge to trigger this interrupt and resets the frequency and duty
+ *   values. The positive edge is always the first expected.
+ *
+ * Input Parameters:
+ *   lower - Pointer to the capture channel lower-half data structure.
+ *
+ * Returned Value:
+ *   Returns OK on success.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE
+static int esp_capture_start(struct cap_lowerhalf_s *lower)
+{
+  struct mcpwm_cap_channel_lowerhalf_s *priv = (
+    struct mcpwm_cap_channel_lowerhalf_s *)lower;
+  mcpwm_hal_context_t *hal = &priv->common->hal;
+
+  DEBUGASSERT(priv != NULL);
+  DEBUGASSERT(priv->common->initialized);
+
+  /* Enable channel and interruption for rising edge */
+
+  mcpwm_ll_capture_enable_channel(hal->dev, priv->channel_id, true);
+  mcpwm_ll_intr_enable(hal->dev,
+                       MCPWM_LL_EVENT_CAPTURE(priv->channel_id),
+                       true);
+  mcpwm_ll_intr_clear_status(hal->dev,
+                             MCPWM_LL_EVENT_CAPTURE(priv->channel_id));
+  mcpwm_ll_capture_enable_posedge(hal->dev, priv->channel_id, true);
+  mcpwm_ll_capture_enable_negedge(hal->dev, priv->channel_id, false);
+
+  /* Reset values of interest */
+
+  priv->freq = 0;
+  priv->duty = 0;
+  priv->isr_count = 0;
+  priv->enabled = true;
+  priv->ready = false;
+
+  cpinfo("Channel enabled: %d\n", priv->channel_id);
+  return OK;
+}
+#endif
+
+/****************************************************************************
+ * Name: esp_capture_stop
+ *
+ * Description:
+ *   This function is a requirement of the upper-half driver. When called,
+ *   disables the capture channel and the interrupt routine associated.
+ *
+ * Input Parameters:
+ *   lower - Pointer to the capture channel lower-half data structure.
+ *
+ * Returned Value:
+ *   Returns OK on success.
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_ESP_MCPWM_CAPTURE
+static int esp_capture_stop(struct cap_lowerhalf_s *lower)
+{
+  struct mcpwm_cap_channel_lowerhalf_s *priv = (
+    struct mcpwm_cap_channel_lowerhalf_s *)lower;
+  mcpwm_hal_context_t *hal = &priv->common->hal;
```

Comment:
Do we need `DEBUGASSERT(priv != NULL);` and similar checks like in `esp_capture_start`?

Reply:
Wouldn't hurt. Just added.